packages/app-builder-lib/src/node-module-collector/hoist.ts

```diff
@@ -0,0 +1,1108 @@
+// copy from https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-nm/sources/hoist.ts
```

Comment:
You can install https://www.npmjs.com/package/@yarnpkg/nm directly so we can track upstream updates to hoisting logic 🙁 
It'll be located here: node_modules/@yarnpkg/nm/lib/hoist.js

Reply:
I considered directly importing it(@yarnpkg/nm), but that would result in importing almost all of the yarnpkg packages. Currently, we only need hoist.ts, which has zero dependencies.

<img width="1235" alt="image" src="https://github.com/user-attachments/assets/fd5f89e7-f8c8-48df-b03d-a24034bbca92">

If you think importing these additional yarnpkg packages isn't a problem, I can change it to a direct import method.

Reply:
Oomph, that's a great callout. I think it'd be safer to track `yarnpkg/nm` upstream though in case the dependency tree logic changes on their side (or they patch a bug)

Curious, can we leverage the result from https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-nm/sources/buildNodeModulesTree.ts `buildNodeModulesTree` since it's in `yarnpkg/nm` instead of the custom logic in packages/app-builder-lib/src/node-module-collector/nodeModulesCollector.ts? Or does `nodeModulesCollector.ts` have additional logic or requirements for us?

Reply:
> Oomph, that's a great callout. I think it'd be safer to track `yarnpkg/nm` upstream though in case the dependency tree logic changes on their side (or they patch a bug)
> 
![image](https://github.com/user-attachments/assets/3315fe12-769f-46d2-8b0e-dec02e17bbbf)
Looking at the historical commits, changes to this file have been infrequent, so we can record each commit in the file header. When there's a new one, we can just copy it. This effort should be minimal.

> Curious, can we leverage the result from https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-nm/sources/buildNodeModulesTree.ts `buildNodeModulesTree` since it's in `yarnpkg/nm` instead of the custom logic in packages/app-builder-lib/src/node-module-collector/nodeModulesCollector.ts? Or does `nodeModulesCollector.ts` have additional logic or requirements for us?

Currently, we can't use buildNodeModulesTree, as it depends on the pnp API, which is only suitable for yarn, not for npm and pnpm. At present, 'yarn list' doesn't return the path for each module, so for the yarn package manager, I'm currently using 'npm list'. The two are compatible, and so far, I haven't encountered any issues.

I'm referencing the approach of pnpm real hoisting and UT, which allows for a relatively unified implementation.
https://github.com/pnpm/pnpm/blob/main/pkg-manager/real-hoist/src/index.ts#L8

https://github.com/yarnpkg/berry/blob/d63d411bcc5adcbffd198b8987c5a14c81eaf669/packages/yarnpkg-nm/tests/hoist.test.ts#L3-L23

Reply:
Should we pull in the full unit test suite for specifically the hoist logic as well then? That way we prevent regressions if we need to make patches on our side.

I'd also recommend adding those links as a comment to the code so we know where it came from and can check back if they make patches upstream 😅

Reply:
> Should we pull in the full unit test suite for specifically the hoist logic as well then? That way we prevent regressions if we need to make patches on our side.
> 
Good suggestion. I have pulled it:
<img width="1080" alt="image" src="https://github.com/user-attachments/assets/a0b2e549-3ab6-4154-a2d6-8c5dbfc94610">

> I'd also recommend adding those links as a comment to the code so we know where it came from and can check back if they make patches upstream 😅
Yes. I have linked both the code links and commit information here.
<img width="516" alt="image" src="https://github.com/user-attachments/assets/fa480680-dcee-4a5c-84be-16b987d16149">